application/src/ext/java/org/opentripplanner/ext/siri/updater/mqtt/MqttEstimatedTimetableSource.java

```diff
@@ -0,0 +1,343 @@
+import com.hivemq.client.mqtt.mqtt5.message.publish.Mqtt5Publish;
+import io.micrometer.core.instrument.FunctionCounter;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.Metrics;
+import jakarta.xml.bind.JAXBException;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+import javax.xml.stream.XMLStreamException;
+import org.entur.siri21.util.SiriXml;
+import org.opentripplanner.updater.trip.siri.updater.AsyncEstimatedTimetableSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import uk.org.siri.siri21.ServiceDelivery;
+import uk.org.siri.siri21.Siri;
+
+/**
+ * This is a realtime updater for trip updates in Siri ET format via MQTT. The updater is primed
+ * (ready for routing requests), when all retained messages in the connected MQTT are processed.
+ * If there are no retained messages, the updater is primed immediately. Live messages (messages
+ * without a retained flag) are always processed, even if the updater is not yet primed.
+ */
+public class MqttEstimatedTimetableSource implements AsyncEstimatedTimetableSource {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MqttEstimatedTimetableSource.class);
+
+  private final MqttSiriETUpdaterParameters parameters;
+
+  private Mqtt5AsyncClient client;
+  private Function<ServiceDelivery, Future<?>> serviceDeliveryConsumer;
+
+  private final BlockingQueue<byte[]> liveMessageQueue = new LinkedBlockingQueue<>();
+  private final BlockingQueue<byte[]> primingMessageQueue = new LinkedBlockingQueue<>();
+  private final ExecutorService primingExecutor;
+  private final ExecutorService liveExecutor;
+
+  private volatile boolean primed = false;
+
+  private Instant connectedAt;
+  private final AtomicLong liveMessageCounter = new AtomicLong();
+  private final AtomicLong primingMessageCounter = new AtomicLong();
+  private final AtomicLong liveMessageSize = new AtomicLong();
+  private final AtomicLong primingMessageSize = new AtomicLong();
+  private final AtomicLong processedLiveMessageCounter = new AtomicLong();
+  private final AtomicLong processedPrimingMessageCounter = new AtomicLong();
+
+  public MqttEstimatedTimetableSource(MqttSiriETUpdaterParameters parameters) {
+    this.parameters = parameters;
+    this.primingExecutor = Executors.newFixedThreadPool(parameters.numberOfPrimingWorkers());
```

Comment:
Configuring the naming of the threads may help when debugging, see
https://github.com/OpenTripPlanner/OpenTripPlanner/blob/5198c5ffef3d2db4f78f36778501333d02ec8444/application/src/main/java/org/opentripplanner/updater/GraphUpdaterManager.java#L79